cocotb/drivers/amba.py

```diff
@@ -220,98 +223,216 @@ class AXI4Slave(BusDriver):
-            _awburst = int(self.bus.AWBURST)
-            _awprot = int(self.bus.AWPROT)
+    @cocotb.coroutine
+    def _write_response(self, awaddr, awlen, awsize,
+                        awburst, awprot, awid, wdata, wid, wstrb):
 
-            burst_length = _awlen + 1
-            bytes_in_beat = self._size_to_bytes_in_beat(_awsize)
+        bresp = 0
+        try:
+            yield self._data_transfer(awaddr, 2**awsize, awlen + 1, awburst, True, wdata, wstrb)
+        except Exception as e:
+            self.log.warning("SAXI Error: {}".format(str(e)))
+            bresp = 2
 
-            word = BinaryValue(n_bits=bytes_in_beat*8, bigEndian=self.big_endian)
+        for i in range(self.bresp_delay-1):
+            yield RisingEdge(self.clock)
 
-            if __debug__:
-                self.log.debug(
-                    "AWADDR  %d\n" % _awaddr +
-                    "AWLEN   %d\n" % _awlen +
-                    "AWSIZE  %d\n" % _awsize +
-                    "AWBURST %d\n" % _awburst +
-                    "BURST_LENGTH %d\n" % burst_length +
-                    "Bytes in beat %d\n" % bytes_in_beat)
+        self.bus.BID <= wid
+        self.bus.BRESP <= bresp
+        self.bus.BVALID <= 1
 
-            burst_count = burst_length
+        yield RisingEdge(self.clock)
+        while not self.bus.BREADY:
+            yield RisingEdge(self.clock)
 
-            yield clock_re
+    @cocotb.coroutine
+    def _data_transfer(self, start_address, num_bytes, burst_length, mode, is_write, data, strbs):
+        """Perform AXI4 Data Transfer
 
-            while True:
-                if self.bus.WVALID.value:
-                    word = self.bus.WDATA.value
-                    word.big_endian = self.big_endian
-                    _burst_diff = burst_length - burst_count
-                    _st = _awaddr + (_burst_diff * bytes_in_beat)  # start
-                    _end = _awaddr + ((_burst_diff + 1) * bytes_in_beat)  # end
-                    self._memory[_st:_end] = array.array('B', word.get_buff())
-                    burst_count -= 1
-                    if burst_count == 0:
-                        break
-                yield clock_re
+        This method is basically the psuedocode from A3.4.2 of the Protocol Specification
+        """
+        # [FIXED, INCR, WRAP]
+        assert mode in [0, 1, 2], "Unsupported Burst Mode: 0x{:1X}".format(mode)
+        assert len(data) == burst_length, "Data supplied doesn't match burst_length"
+        assert len(strbs) == burst_length, "Data supplied doesn't match burst_length"
+
+        addr = start_address
+        aligned_address = (addr // num_bytes) * num_bytes
+        aligned = aligned_address == addr
+        dtsize = num_bytes * burst_length
+        lower_wrap_boundary = (start_address // (num_bytes * burst_length)) \
+                               * (num_bytes * burst_length)
+
+        if mode == 2:  # WRAP
+            lower_wrap_boundary = (addr // dtsize) * dtsize
```

Comment:
This comes up enough I'd recommend a  helper
```
def _floor_to(x, divisor):
    """ round x down to the nearest multiple of divisor """
    return x // divisor * divisor
```

Reply:
I'd agree, but for the resemblance to the pseudocode in the spec.  Is it worth maintaining? or should I boil it down to something more pythonic?

Reply:
Matching the spec is a fair argument